---
ms.date: 09/13/2016
ms.topic: reference
title: 终止错误
description: 终止错误
ms.openlocfilehash: b7c9b949a654f10a0421ea69dfe955c8dfb5627e
ms.sourcegitcommit: ba7315a496986451cfc1296b659d73ea2373d3f0
ms.translationtype: MT
ms.contentlocale: zh-CN
ms.lasthandoff: 12/10/2020
ms.locfileid: "92652548"
---
# <a name="terminating-errors"></a><span data-ttu-id="30d5c-103">终止错误</span><span class="sxs-lookup"><span data-stu-id="30d5c-103">Terminating Errors</span></span>

<span data-ttu-id="30d5c-104">本主题讨论用于报告终止错误的方法。</span><span class="sxs-lookup"><span data-stu-id="30d5c-104">This topic discusses the method used to report terminating errors.</span></span> <span data-ttu-id="30d5c-105">它还讨论了如何从 cmdlet 内调用方法，并讨论了调用方法时 Windows PowerShell 运行时可能返回的异常。</span><span class="sxs-lookup"><span data-stu-id="30d5c-105">It also discusses how to call the method from within the cmdlet, and it discusses the exceptions that can be returned by the Windows PowerShell runtime when the method is called.</span></span>

<span data-ttu-id="30d5c-106">当发生终止错误时，此 cmdlet 应通过调用 [Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) 方法报告错误消息。</span><span class="sxs-lookup"><span data-stu-id="30d5c-106">When a terminating error occurs, the cmdlet should report the error by calling the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method.</span></span> <span data-ttu-id="30d5c-107">此方法允许 cmdlet 发送描述引起终止错误的条件的错误记录。</span><span class="sxs-lookup"><span data-stu-id="30d5c-107">This method allows the cmdlet to send an error record that describes the condition that caused the terminating error.</span></span> <span data-ttu-id="30d5c-108">有关错误记录的详细信息，请参阅 [Windows PowerShell 错误记录](./windows-powershell-error-records.md)。</span><span class="sxs-lookup"><span data-stu-id="30d5c-108">For more information about error records, see [Windows PowerShell Error Records](./windows-powershell-error-records.md).</span></span>

<span data-ttu-id="30d5c-109">如果调用了 [Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) 方法，则 Windows PowerShell 运行时将永久停止执行管道，并引发 [Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) 的异常情况下的操作。</span><span class="sxs-lookup"><span data-stu-id="30d5c-109">When the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method is called, the  Windows PowerShell runtime permanently stops the execution of the pipeline and throws a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span> <span data-ttu-id="30d5c-110">任何后续尝试调用[WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject)、WriteError 或其他几个 api，都将导致这些调用引发[Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException)异常的情况下进行调用，从而引发[](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)异常。</span><span class="sxs-lookup"><span data-stu-id="30d5c-110">Any subsequent attempts to call [System.Management.Automation.Cmdlet.WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError), or several other APIs causes those calls to throw a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span>

<span data-ttu-id="30d5c-111">如果管道中的另一个 cmdlet 报告终止错误（如果用户已请求停止管道），或者如果由于任何原因已暂停管道，则也会发生 [Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) 异常。</span><span class="sxs-lookup"><span data-stu-id="30d5c-111">The [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception can also occur if another cmdlet in the pipeline reports a terminating error, if the user has asked to stop the pipeline, or if the pipeline has been halted before completion for any reason.</span></span> <span data-ttu-id="30d5c-112">Cmdlet 不需要捕获 [Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) 异常，除非它必须清除打开的资源或其内部状态。</span><span class="sxs-lookup"><span data-stu-id="30d5c-112">The cmdlet does not need to catch the [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception unless it must clean up open resources or its internal state.</span></span>

<span data-ttu-id="30d5c-113">在报告终止错误之前，cmdlet 可以写入任意数量的输出对象或非终止错误。</span><span class="sxs-lookup"><span data-stu-id="30d5c-113">Cmdlets can write any number of output objects or non-terminating errors before reporting a terminating error.</span></span> <span data-ttu-id="30d5c-114">但是，终止错误会永久停止管道，并且不会报告其他输出、终止错误或非终止错误。</span><span class="sxs-lookup"><span data-stu-id="30d5c-114">However, the terminating error permanently stops the pipeline, and no further output, terminating errors, or non-terminating errors can be reported.</span></span>

<span data-ttu-id="30d5c-115">Cmdlet 只能从名为 BeginProcessing、 [ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)[或](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)Throwterminatingerror 输入处理方法的线程调用[](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing) [\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)类型的调用程序中的一个类型的调用程序中的调用程序中的方法的类型为。</span><span class="sxs-lookup"><span data-stu-id="30d5c-115">Cmdlets can call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) only from the thread that called the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) input processing method.</span></span> <span data-ttu-id="30d5c-116">不要尝试从另一个线程调用 [Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) 或 [WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) 的其他线程中的其他线程。</span><span class="sxs-lookup"><span data-stu-id="30d5c-116">Do not attempt to call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) or [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) from another thread.</span></span> <span data-ttu-id="30d5c-117">而必须将错误发回到主线程。</span><span class="sxs-lookup"><span data-stu-id="30d5c-117">Instead, errors must be communicated back to the main thread.</span></span>

<span data-ttu-id="30d5c-118">Cmdlet 可在其实现[BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)、 [ProcessRecord 或](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)方法的实现中引发异常，这种情况可能会引发异常的情况下引发异常的情况下可能引发[异常。](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)</span><span class="sxs-lookup"><span data-stu-id="30d5c-118">It is possible for a cmdlet to throw an exception in its implementation of the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) method.</span></span> <span data-ttu-id="30d5c-119">从这些方法引发的任何异常 (除了几个停止 Windows PowerShell 主机) 的严重错误情况外，还会被解释为一个终止错误，该错误将停止管道，而不是整个 Windows PowerShell。</span><span class="sxs-lookup"><span data-stu-id="30d5c-119">Any exception thrown from these methods (except for a few severe error conditions that stop the Windows PowerShell host) is interpreted as a terminating error which stops the pipeline, but not Windows PowerShell as a whole.</span></span> <span data-ttu-id="30d5c-120"> (这仅适用于主 cmdlet 线程。</span><span class="sxs-lookup"><span data-stu-id="30d5c-120">(This applies only to the main cmdlet thread.</span></span> <span data-ttu-id="30d5c-121">通常情况下，在由 cmdlet 生成的线程中未捕获到异常。 ) 建议使用 [Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) 而不是引发异常，因为错误记录提供了有关错误情况的其他信息，这对最终用户非常有用。</span><span class="sxs-lookup"><span data-stu-id="30d5c-121">Uncaught exceptions in threads spawned by the cmdlet, in general, halt the Windows PowerShell host.) We recommend that you use [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) rather than throwing an exception because the error record provides additional information about the error condition, which is useful to the end-user.</span></span> <span data-ttu-id="30d5c-122">Cmdlet 应遵循托管代码准则，以避免捕获和处理)  (`catch (Exception e)` 的所有异常。</span><span class="sxs-lookup"><span data-stu-id="30d5c-122">Cmdlets should honor the managed code guideline against catching and handling all exceptions (`catch (Exception e)`).</span></span> <span data-ttu-id="30d5c-123">仅将已知类型和预期类型的异常转换为错误记录。</span><span class="sxs-lookup"><span data-stu-id="30d5c-123">Convert only exceptions of known and expected types into error records.</span></span>

## <a name="see-also"></a><span data-ttu-id="30d5c-124">另请参阅</span><span class="sxs-lookup"><span data-stu-id="30d5c-124">See Also</span></span>

[<span data-ttu-id="30d5c-125">"BeginProcessing"。</span><span class="sxs-lookup"><span data-stu-id="30d5c-125">System.Management.Automation.Cmdlet.BeginProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)

[<span data-ttu-id="30d5c-126">System.web. EndProcessing. EndProcessing</span><span class="sxs-lookup"><span data-stu-id="30d5c-126">System.Management.Automation.Cmdlet.EndProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)

[<span data-ttu-id="30d5c-127">"ProcessRecord"。</span><span class="sxs-lookup"><span data-stu-id="30d5c-127">System.Management.Automation.Cmdlet.ProcessRecord</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)

[<span data-ttu-id="30d5c-128">System.web. Pipelinestoppedexception</span><span class="sxs-lookup"><span data-stu-id="30d5c-128">System.Management.Automation.Pipelinestoppedexception</span></span>](/dotnet/api/System.Management.Automation.PipelineStoppedException)

[<span data-ttu-id="30d5c-129">Throwterminatingerror \* 的 \*</span><span class="sxs-lookup"><span data-stu-id="30d5c-129">System.Management.Automation.Cmdlet.Throwterminatingerror\*</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)

[<span data-ttu-id="30d5c-130">"WriteError"。</span><span class="sxs-lookup"><span data-stu-id="30d5c-130">System.Management.Automation.Cmdlet.WriteError</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)

[<span data-ttu-id="30d5c-131">Windows PowerShell 错误记录</span><span class="sxs-lookup"><span data-stu-id="30d5c-131">Windows PowerShell Error Records</span></span>](./windows-powershell-error-records.md)

[<span data-ttu-id="30d5c-132">编写 Windows PowerShell Cmdlet</span><span class="sxs-lookup"><span data-stu-id="30d5c-132">Writing a Windows PowerShell Cmdlet</span></span>](./writing-a-windows-powershell-cmdlet.md)
